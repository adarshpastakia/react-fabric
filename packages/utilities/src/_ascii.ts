/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* istanbul ignore file */

import { isEmpty } from "./_isType";

/** ***************** latin-ascii conversions *******************/
const conversions = {
  AE: "Æ|Ǽ",
  Ae: "Ä",
  ae: "ä|æ|ǽ",
  OE: "Œ",
  Oe: "Ö",
  oe: "ö|œ",
  Ue: "Ü",
  ue: "ü",
  IJ: "Ĳ",
  ij: "ĳ",
  ss: "ß",
  A: "À|Á|Â|Ã|Ä|Å|Ǻ|Ā|Ă|Ą|Ǎ",
  a: "à|á|â|ã|å|ǻ|ā|ă|ą|ǎ|ª",
  C: "Ç|Ć|Ĉ|Ċ|Č",
  c: "ç|ć|ĉ|ċ|č",
  D: "Ð|Ď|Đ",
  d: "ð|ď|đ",
  E: "È|É|Ê|Ë|Ē|Ĕ|Ė|Ę|Ě",
  e: "è|é|ê|ë|ē|ĕ|ė|ę|ě",
  G: "Ĝ|Ğ|Ġ|Ģ",
  g: "ĝ|ğ|ġ|ģ",
  H: "Ĥ|Ħ",
  h: "ĥ|ħ",
  I: "Ì|Í|Î|Ï|Ĩ|Ī|Ĭ|Ǐ|Į|İ",
  i: "ì|í|î|ï|ĩ|ī|ĭ|ǐ|į|ı",
  J: "Ĵ",
  j: "ĵ",
  K: "Ķ",
  k: "ķ",
  L: "Ĺ|Ļ|Ľ|Ŀ|Ł",
  l: "ĺ|ļ|ľ|ŀ|ł",
  N: "Ñ|Ń|Ņ|Ň",
  n: "ñ|ń|ņ|ň|ŉ",
  O: "Ò|Ó|Ô|Õ|Ō|Ŏ|Ǒ|Ő|Ơ|Ø|Ǿ",
  o: "ò|ó|ô|õ|ō|ŏ|ǒ|ő|ơ|ø|ǿ|º",
  R: "Ŕ|Ŗ|Ř",
  r: "ŕ|ŗ|ř",
  S: "Ś|Ŝ|Ş|Š",
  s: "ś|ŝ|ş|š|ſ",
  T: "Ţ|Ť|Ŧ",
  t: "ţ|ť|ŧ",
  U: "Ù|Ú|Û|Ũ|Ū|Ŭ|Ů|Ű|Ų|Ư|Ǔ|Ǖ|Ǘ|Ǚ|Ǜ",
  u: "ù|ú|û|ũ|ū|ŭ|ů|ű|ų|ư|ǔ|ǖ|ǘ|ǚ|ǜ",
  Y: "Ý|Ÿ|Ŷ",
  y: "ý|ÿ|ŷ",
  W: "Ŵ",
  w: "ŵ",
  Z: "Ź|Ż|Ž",
  z: "ź|ż|ž",
  f: "ƒ",
};

/**
 * latin to ascii converter for convenient searching
 * @internal
 */
export const ascii = (str: string) => {
  if (isEmpty(str)) return "";
  Object.entries(conversions).forEach(([a, l]) => {
    const re = new RegExp(l, "g");
    str = `${str}`.replace(re, a);
  });
  return str;
};
