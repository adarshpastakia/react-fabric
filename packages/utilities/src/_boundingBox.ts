/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { isEmpty } from "./_isType";

const _getBox = (
  box: string | [number, number, number, number],
): [number, number, number, number] => {
  let boundingBox: any = box;
  if (isEmpty(box)) return [0, 0, 0, 0];
  if (typeof box === "string") {
    boundingBox = box.split(",");
  }
  if (Array.isArray(boundingBox)) {
    boundingBox = boundingBox.map((b) => parseInt(b, 10));
  }
  return boundingBox;
};

/**
 * box "x,y,w,h" | [x,y,w,h]
 * returns canvas box [x,y,w,h]
 */
export const getBox = (
  box: string | [number, number, number, number] = "0,0,0,0",
) => {
  const boundingBox = _getBox(box);

  let [x, y, w, h] = boundingBox;
  w = x < 0 ? w + x : w;
  h = y < 0 ? h + y : h;
  x = x > 0 ? x : 0;
  y = y > 0 ? y : 0;
  return [x, y, w, h] as [number, number, number, number];
};

/**
 * box "x1,y1,x2,y2" | [x1,y1,x2,y2]
 * returns canvas box [x,y,w,h]
 */
export const getBoundingBox = (
  box: string | [number, number, number, number] = "0,0,0,0",
) => {
  const boundingBox = _getBox(box);

  let [x1, y1, x2, y2] = boundingBox;
  const w = x2 - x1;
  const h = y2 - y1;
  x1 = x1 > 0 ? x1 : 0;
  y1 = y1 > 0 ? y1 : 0;
  return [x1, y1, w, h] as [number, number, number, number];
};
