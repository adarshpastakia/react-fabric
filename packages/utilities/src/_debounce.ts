/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* istanbul ignore file */

/**
 * Debounce function to limit the rate at which a function can fire.
 * It will delay the execution of the function until after a specified timeout
 * has elapsed since the last time the function was invoked.
 *
 * @param fn - The function to debounce.
 * @param timeout - The number of milliseconds to wait before invoking the function.
 * @returns A debounced version of the function that can be called with any arguments.
 *          The debounced function also has a `cancel` method to clear the timeout.
 */
export const debounce = (fn: (...rest: any) => any, timeout = 500) => {
  /** ***************** create debounce timer id using callback hash *******************/
  let _timer: any;

  /** ***************** return debounced function *******************/
  const cb = (...args: any) => {
    clearTimeout(_timer);
    if (typeof window !== "undefined")
      _timer = window.setTimeout(() => fn?.(...args), timeout);
  };
  cb.cancel = () => clearTimeout(_timer);
  return cb;
};
