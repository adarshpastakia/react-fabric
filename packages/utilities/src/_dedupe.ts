/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { isEmpty } from "./_isType";

export const dedupe = (list: any[], key?: string) => {
  const filtered = list.filter((i) => !isEmpty(i));
  if (!key) {
    return Array.from(new Set(filtered).values());
  }

  return Array.from(
    new Map(filtered.map((obj) => [obj[key] ?? obj, obj])).values(),
  );
};

export const flatten = (list: any[]) => {
  return list.flat(5).filter((i) => !isEmpty(i));
};

export const flattenAndDedupe = (list: any[] = [], key?: string) => {
  return dedupe(flatten(list), key);
};

export const groupBy = <T extends KeyValue>(
  list: T[],
  prop: keyof T,
  missingKey?: string,
): KeyValue<T[]> => {
  return list.reduce((r: AnyObject, i) => {
    const key = i[prop] ?? missingKey;
    if (!r[key]) r[key] = [];
    r[key].push(i);
    return r;
  }, {});
};
