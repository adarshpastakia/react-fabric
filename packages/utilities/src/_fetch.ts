/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* istanbul ignore file */

import { isString } from "./_isType";

declare type METHODS = "get" | "post" | "put" | "delete";
declare interface FetchOptions {
  body?: KeyValue | FormData;
  headers?: KeyValue;
  signal?: string | AbortSignal;
}

const _signals = new Map<string, AbortController>();

/**
 * Fetch helper
 * @param method
 * @param url
 * @param options.headers
 * @param options.body
 * @param options.signal
 * @internal
 */
export const _fetch = async (
  method: METHODS,
  url: string,
  { headers = {}, body, signal }: FetchOptions,
) => {
  const options: KeyValue = {
    method,
  };

  headers.accept = "application/json";
  if (method === "post" || method === "put") {
    options.body = body;
    /** ***************** set body to json when not FormData *******************/
    if (!(body instanceof FormData)) {
      headers["content-type"] = "application/json";
      options.body = JSON.stringify(body);
    }
  }

  /** ***************** reset abort signal *******************/
  if (signal && isString(signal)) {
    if (_signals.has(signal)) {
      const _s = _signals.get(signal);
      _s?.abort();
      _signals.delete(signal);
    }
    const ac = new AbortController();
    options.signal = ac.signal;
    _signals.set(signal, ac);
  } else if (signal instanceof AbortSignal) {
    options.signal = signal;
  }

  return await fetch(url, { ...options, headers })
    .then(async (resp) => {
      if (resp.status >= 200 && resp.status < 300) {
        return await resp.json();
      }
      // eslint-disable-next-line @typescript-eslint/no-throw-literal
      throw { status: resp.status, code: -1, message: resp.statusText };
    })
    .then((resp) => {
      /** ***************** check for possible error code propeties in response *******************/
      const code = resp.error_code ?? resp.errorCode ?? resp.error;
      const message = resp.error_message ?? resp.errorMessage ?? resp.message;
      if (resp.status === "error" || code) {
        // eslint-disable-next-line @typescript-eslint/no-throw-literal
        throw { status: 500, code, message };
      }
      return resp;
    });
};

/** ***************** attach isAborted helper to fetch method *******************/
_fetch.isAborted = ({ name }: Error) => name === "AbortError";

/** ***************** attach abort helper to fetch method *******************/
_fetch.abort = (signal: string) => {
  if (_signals.has(signal)) {
    const _s = _signals.get(signal);
    _s?.abort();
    _signals.delete(signal);
  }
};
