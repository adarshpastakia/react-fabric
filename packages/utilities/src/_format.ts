/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { differenceInDays, format, parseISO } from "date-fns";
import {
  parsePhoneNumberFromString,
  type CountryCode,
} from "libphonenumber-js";
import numeral from "numeral";
import Countries from "./_countries";
import { isEmpty, isNil } from "./_isType";

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace Format {
  /** ***************** get phone number object using `libphonenumber-js` *******************/
  const getPhone = (value: string, country: string) => {
    const phone =
      value.match(/^[0+]/) != null ? value.replace(/^00/, "+") : `+${value}`;
    const number = parsePhoneNumberFromString(phone, country as CountryCode);
    /* istanbul ignore next */
    return (
      number ?? {
        country: "",
        formatInternational: () => value,
      }
    );
  };

  /** ***************** common number format method using `numeral` *******************/
  const numberFormat = (number?: string | number, format = "0,0[.]00a") => {
    if (isNil(number)) {
      return "";
    }
    const prefix = `${number}`.startsWith("+") ? "+" : "";
    return prefix + numeral(number).format(format);
  };

  const makeDuration = (value: number, fmt = "HH:mm:ss.SSS") => {
    if (!isNaN(value)) {
      const days = differenceInDays(value, 0);
      const prefix = days > 0 ? `${days}d:` : "";
      let time = format(
        parseISO(new Date(value).toISOString().replace("Z", "")),
        fmt,
      );
      if (!prefix) time = time.replace(/^00:/, "");
      return `${prefix}${time}`;
    }
  };

  /** ***************** format phone number using `libphonenumber-js` *******************/
  export const phone = (value?: string, country = "AE", useCss = false) => {
    if (isEmpty(value) || value.length < 6 || !/^[\d+\s\-()]+$/.test(value))
      return value;
    const phone = getPhone(value, country);
    /* istanbul ignore next */
    return `${
      useCss
        ? `<span class="flag ${phone.country}"></span>`
        : Countries.emoji(phone.country ?? "")
    } ${phone.formatInternational()}`;
  };
  /** ***************** format whole number using `numeral` *******************/
  export const number = (number?: string | number, format?: string) => {
    return numberFormat(number, format);
  };
  /** ***************** format bytes using `numeral` *******************/
  export const bytes = (number?: string | number) => {
    return numberFormat(number, "0,0[.]00b");
  };
  /** ***************** format percentage using `numeral` *******************/
  export const percent = (number?: string | number) => {
    return numberFormat(number, "0,0[.]00%");
  };
  /** ***************** format date using `date-fns` *******************/
  export const date = (date?: Date | string | number, fmt = "d-M-y") => {
    return date ? format(new Date(date), fmt) : "";
  };
  /** ***************** format time duration from total seconds/milliseconds *******************/
  /**
   * Format time duration with milliseconds
   * @param number
   * @param isFraction
   * @returns
   */
  export const duration = (number?: string | number, isFraction?: boolean) => {
    if (isNil(number)) {
      return "00:00.000";
    }
    if (number === Infinity) {
      return "∞";
    }
    const value = parseFloat(`${number}`) * (isFraction ? 1000 : 1);
    if (!isNaN(value)) {
      return makeDuration(value);
    }
    return "00:00.000";
  };
  /**
   * Format time duration without milliseconds
   * @param number
   * @param isFraction
   * @returns
   */
  export const durationSeconds = (
    number?: string | number,
    isFraction?: boolean,
  ) => {
    if (isNil(number)) {
      return "00:00";
    }
    if (number === Infinity) {
      return "∞";
    }
    const value = parseFloat(`${number}`) * (isFraction ? 1000 : 1);
    if (!isNaN(value)) {
      return makeDuration(value, "HH:mm:ss");
    }
    return "00:00";
  };
}
