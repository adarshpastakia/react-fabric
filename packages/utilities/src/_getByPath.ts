/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { isArray, isEmpty, isNil, isObject } from "./_isType";

const _get = (hit: any, field: string): any => {
  /** ***************** return if object is empty *******************/
  /* istanbul ignore next */
  if (isNil(hit)) {
    return undefined;
  }
  /** ***************** return object if property is empty *******************/
  /* istanbul ignore next */
  if (isNil(field)) {
    return hit;
  }

  if (isObject(hit)) {
    /** ***************** return value if object.property is not empty *******************/
    if (!isNil(hit[field])) {
      return hit[field];
    }

    /** ***************** if property name contains `.` separator split to multiple property names *******************/
    if (field.includes(".")) {
      let currentValue: any = hit;

      const path = field.split(".");
      for (let i = path.length - 1; i >= 0; i--) {
        /** ***************** when property name is `a.b.c` iterate until property found in object. eg. `a`, `a.b` *******************/
        const outerPath = path.slice(0, i as any).join(".");

        /** ***************** join rest of property list *******************/
        const innerPath = path.slice(i as any).join(".");

        /** ***************** if current outer path found in object *******************/
        if (outerPath in currentValue) {
          currentValue = currentValue[outerPath];

          /** ***************** check for type array *******************/
          if (isArray(currentValue)) {
            /** ***************** return value if remaining path found *******************/
            if (currentValue[innerPath as any]) {
              return currentValue[innerPath as any];
            }
            /** ***************** return array map of inner propeties *******************/
            return currentValue
              .map((val) => _get(val, innerPath))
              .filter((val) => !isEmpty(val));
          } else {
            /** ***************** return value of remaining path *******************/
            return _get(currentValue, innerPath);
          }
        }
      }
    }
  }

  return undefined;
};

/**
 * get property value by path
 * @internal
 */
export const getByPath = <T = any>(
  obj: KeyValue,
  path: string,
  defaultValue: any = undefined,
): T => {
  /** ***************** if empty object/path return defaultValue *******************/
  if (isEmpty(obj) || isEmpty(path)) {
    return defaultValue;
  }
  const value = _get(obj, path);
  if (!isNil(value)) {
    return value;
  }
  return defaultValue;
};

/**
 * get first available value in list
 * @internal
 */
export const getValue = (...args: any[]) => {
  return args.find((a) => !isEmpty(a));
};
