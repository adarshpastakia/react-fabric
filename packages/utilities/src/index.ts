/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* istanbul ignore file */

export * from "./_ascii";
export * from "./_boundingBox";
export { getBoundingBox, getBox } from "./_boundingBox";
export { default as Countries } from "./_countries";
export type { Country } from "./_countries";
export * from "./_debounce";
export * from "./_dedupe";
export { _fetch as fetch } from "./_fetch";
export { default as FileUtil } from "./_fileType";
export * from "./_format";
export * from "./_getByPath";
export * from "./_hash";
export * from "./_interpolate";
export * from "./_isEqual";
export * from "./_isType";
export * from "./_mergeRefs";
export * from "./_tokenize";
export { getImageColorset } from "./getImageColorset";
export { useLogger } from "./useLogger";
export * from "./yup";

export const EMPTY_ARRAY = [];
Object.freeze(EMPTY_ARRAY);

export const EMPTY_OBJECT = {};
Object.freeze(EMPTY_OBJECT);
