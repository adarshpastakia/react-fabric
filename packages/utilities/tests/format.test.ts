/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { FileUtil, Format } from "../src";

describe("format test", () => {
  it("should format number", (done) => {
    expect(Format.number()).toBe("");
    expect(Format.number(9.9)).toBe("9.90");
    expect(Format.number(9990)).toBe("9.99k");
    expect(Format.number("-9.81")).toBe("-9.81");
    expect(Format.number("+9.24")).toBe("+9.24");
    done();
  });

  it("should format bytes", (done) => {
    expect(Format.bytes()).toBe("");
    expect(Format.bytes(1000)).toBe("1KB");
    expect(Format.bytes(1000 * 1000)).toBe("1MB");
    expect(Format.bytes(1000 * 1000 * 1000)).toBe("1GB");
    done();
  });

  it("should format percentage", (done) => {
    expect(Format.percent()).toBe("");
    expect(Format.percent(0.24)).toBe("24%");
    expect(Format.percent(0.2418)).toBe("24.18%");
    done();
  });

  it("should format date", (done) => {
    expect(Format.date()).toBe("");
    expect(Format.date("2020-01-01")).toBe("1-1-2020");
    expect(Format.date("2020-01-01", "dd-MMM-y")).toBe("01-Jan-2020");
    done();
  });

  it("should format duration", (done) => {
    expect(Format.duration()).toBe("00:00.000");
    expect(Format.duration("test")).toBe("00:00.000");
    expect(Format.duration(Infinity)).toBe("∞");
    expect(Format.duration(86400000 + 14400000 + 1740000 + 18000 + 189)).toBe(
      "1d:04:29:18.189",
    );
    expect(Format.duration(60 * 24 + 3 + 0.128, true)).toBe("24:03.128");
    done();
  });

  it("should format duration seconds", (done) => {
    expect(Format.durationSeconds()).toBe("00:00");
    expect(Format.durationSeconds("test")).toBe("00:00");
    expect(Format.durationSeconds(Infinity)).toBe("∞");
    expect(Format.durationSeconds(1000 * 60 * 24 + 3000 + 128)).toBe("24:03");
    expect(Format.durationSeconds(60 * 24 + 3 + 0.128, true)).toBe("24:03");
    done();
  });

  it("should format phone", (done) => {
    expect(Format.phone()).toBe(undefined);
    expect(Format.phone("99123")).toMatchSnapshot();
    expect(Format.phone("badphone")).toMatchSnapshot();
    expect(Format.phone("971502491824")).toMatchSnapshot();
    expect(Format.phone("00971502491824")).toMatchSnapshot();
    expect(Format.phone("00971502491824", "AE", true)).toMatchSnapshot();
    done();
  });

  it("get mime type", (done) => {
    expect(FileUtil.mime("application/pdf")).toBe("PDF");
    expect(FileUtil.mime("image/xyz")).toBe("image (xyz)");
    done();
  });
});
