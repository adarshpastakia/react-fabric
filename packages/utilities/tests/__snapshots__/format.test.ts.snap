// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format test should format phone 1`] = `"99123"`;

exports[`format test should format phone 2`] = `"badphone"`;

exports[`format test should format phone 3`] = `"ðŸ‡¦ðŸ‡ª +971 50 249 1824"`;

exports[`format test should format phone 4`] = `"ðŸ‡¦ðŸ‡ª +971 50 249 1824"`;

exports[`format test should format phone 5`] = `"<span class="flag AE"></span> +971 50 249 1824"`;
